net: openthread: allow to configure CSL debugging

edmont opened the pull request:
Add new `IEEE802154_CSL_DEBUG` Kconfig option that, when enabled, prevents the radio to sleep after a delayed reception is finished.

This allows to debug CSL timing issues due to accuracy drifts in communications between receiver and transmitter.

jciupis commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -1019,7 +1019,9 @@ void nrf_802154_receive_failed(nrf_802154_rx_error_t error, uint32_t id)
 		__ASSERT_NO_MSG(nrf5_data.event_handler);
+#if !defined(CONFIG_IEEE802154_CSL_DEBUG)
 		nrf5_data.event_handler(dev, IEEE802154_EVENT_SLEEP, NULL);
+#endif
```
With this change in place when `CONFIG_IEEE802154_CSL_DEBUG` is set the 802.15.4 RD will indeed stay in receive state, but with `id == DRX_SLOT_RX`. That might lead to errors where failure notifications related to ordinary, indefinite idle listening will be reported with an incorrect ID. I'd suggest explicitly switching to receive state when this debug option is selected. This probably needs to be handled through the higher layer similarly to the sleep request above.

edmont replied:
The main reason to notify the higher layer would be for OT to switch to receive state. Another side effect would be that channel would be switched to PAN channel, which might be different from CSL channel.

Since the main purpose of this configuration option is to debug CSL timing issues, we shouldn't be worried about modifying upper layer state, and actually when delayed reception is used (`otPlatRadioReceiveAt`) the sleep state will remain unchanged.

So, wouldn't a call to `nrf_802154_receive` be enough here to clean the `id` value?

jciupis replied:
> Another side effect would be that channel would be switched to PAN channel, which might be different from CSL channel.

I thought that was desirable. You're correct, the channel would be switched.

> So, wouldn't a call to `nrf_802154_receive` be enough here to clean the `id` value?

Calling 802.15.4 radio driver API directly from its callout will lead to a deadlock in serialization.

If it's a debug option meant primarily for manual testing & investigating and proper handling the notification in OT would remove the very point of this change then in my opinion it is fine to leave the code as it is now. In that case please add a comment here that explains the handling is omitted on purpose, and what consequence it might have to the application (incorrect ID in notifications).

rlubos commented on drivers/ieee802154/Kconfig:
```diff
@@ -90,6 +90,11 @@ config IEEE802154_CSL_ENDPOINT
 
+config IEEE802154_CSL_DEBUG
+	bool "Support for CSL debugging"
+	help
```
`depends on IEEE802154_CSL_ENDPOINT`?

edmont committed:
net: openthread: allow to configure CSL debugging

Add new `IEEE802154_CSL_DEBUG` Kconfig option that, when enabled,
prevents the radio to sleep after a delayed reception is finished.

This allows to debug CSL timing issues due to accuracy drifts in
communications between receiver and transmitter.

Signed-off-by: Eduardo Montoya <eduardo.montoya@nordicsemi.no>
